### ISSUE TITLE
Feedback: How to access records with unreplicated changes?

### ISSUE THREAD

**Issue by MrChadMWood** [pubkey/rxdb#6257]
I am looking for a way to check on pending pushes. Essentially, track in real-time (on the client-side) which how many documents are needing to be replicated to the server-side still. 

I want to implement a counter with this functionality, so the user knows if some of their work is still client-side only (and how much). I.e., a user makes a change to a document, and they see the counter increment by one; once the document replicates to the server, the counter simply decrements. Aside from that, I also want to implement the option to download the unreplicated documents as a .json file. 

Simply put, I'm just looking for:
* A way for the client to know how many docs still need to be pushed to server
* A way to know which docs those are

The idea is that users can see if the counter isn't decrementing, and in that case they know something is potentially wrong. They can download their work as a last resort, if my app has any issues. I don't want to be in a predicament where the data is lost simply because they had no options to download it from their browser.

Can it be done in a way that works regardless of how the change was made? Or, do I need to create observables for each way of modifying the collection? Thank you for any feedback!

**Comment by pubkey** [comment]
For the unreplicated pulls, your server has to tell the client the number.
For pushes there is no official feature. But you could get the push-checkpoint from the replication metadata and then run a query against the storageInstance of the collection.

**Comment by MrChadMWood** [comment]
I am only looking to check on pushes. Basically, a counter will be displayed to the user to let them know how many offline documents still need to be replicated to the server. Any document modifications would increment the counter, and a successful push would decrement. If a user clicks the counter, I want the currently un-replicated docs to be downloaded as JSON.

I will try the approach you've given. Where's the push-checkpoint located on the `replicationState` object?

**Comment by MrChadMWood** [comment]
@pubkey Just following up, how should the `push-checkpoint` be accessed?

**Comment by pubkey** [comment]
This is not documented. You can check the replication internals to fetch the checkpoint.

**Comment by MrChadMWood** [comment]
Thanks that seems like it is what I need. So, I can get the checkpoint doc, then query the collection for any documents with `updated_at` after the checkpoint doc. I guess I can also subscribe to the query, so that my components can react to its changes. This sounds great!

How do you access the [function](https://unpkg.com/browse/rxdb@14.17.1/src/replication-protocol/checkpoint.ts)?

I did try:
```
import { getLastCheckpointDoc } from 'rxdb/replication-protocol/checkpoint';

// output
Failed to compile.

Module not found: Error: Package path ./replication-protocol/checkpoint is not exported from package /app/node_modules/rxdb (see exports field in /app/node_modules/rxdb/package.json)
ERROR in ./src/hooks/useDatabase.jsx 6:0-76
Module not found: Error: Package path ./replication-protocol/checkpoint is not exported from package /app/node_modules/rxdb (see exports field in /app/node_modules/rxdb/package.json)
```

**Comment by MrChadMWood** [comment]
How do you access the [function](https://unpkg.com/browse/rxdb@14.17.1/src/replication-protocol/checkpoint.ts)?

I did try:
```
import { getLastCheckpointDoc } from 'rxdb/replication-protocol/checkpoint';
// Same result with: 'rxdb/src/replication-protocol/checkpoint'

// output
Failed to compile.

Module not found: Error: Package path ./replication-protocol/checkpoint is not exported from package /app/node_modules/rxdb (see exports field in /app/node_modules/rxdb/package.json)
ERROR in ./src/hooks/useDatabase.jsx 6:0-76
Module not found: Error: Package path ./replication-protocol/checkpoint is not exported from package /app/node_modules/rxdb (see exports field in /app/node_modules/rxdb/package.json)
```

**Comment by MrChadMWood** [comment]
@pubkey I'm not able to get a response with direction set to "up"

```
14:29:33.033 await getLastCheckpointDoc(replicationState.internalReplicationState, 'up')
14:29:33.045 undefined 
```

Using "down" works though. Any idea what might cause this?

**Comment by pubkey** [comment]
Maybe there is not checkpoint stored yet when the replication is started?

**Comment by MrChadMWood** [comment]
I had already pushed some docs to the backend, which I assume is what triggers a checkpoint creation. 

Do you know if something like this can work instead?
```
// Subscribes to sent$ to keep last pushed doc as custom /push checkpoint
replicationState.sent$.subscribe(async doc => {
  replicationState.collection.upsertLocal('last-push', { doc: doc });
});
```

I think the only need here is that RxDB would `/push` docs out in the same order that they were changed. What do you think, and does the queue even work that way?

**Comment by MrChadMWood** [comment]
@pubkey Any chance you;d be able to show a minimal example of observing a query containing unreplicated documents on the client side (as in, unreplicated from client to server)? I've made some attempts, but this is a tricky one especially as there is no formal documentation on how to do this.

Attempt 1: I utilized the `getLastCheckpointDoc` but the `'up'` direction continues producing an `undefined` result.
```
14:29:33.033 await getLastCheckpointDoc(replicationState.internalReplicationState, 'up')
14:29:33.045 undefined 
```
To troubleshoot this, I tried making some `/push` replications and checking only after they'd completed. The result was still `undefined`.

Attempt 2: I tried synchronizing the schema's local document ( `last-push`) with the `replicationState.sent$.subscribe` observable, then I observed changes to that local document while subscribed to a query for documents with a newer `updatedAt` value. Whenever the local document `last-push` value changed, I unsubscribed to update the query with the new docs `updatedAt` and subscribed to this new query. I wasn't able to get consistent results with this approach and haven't been able to diagnose why just yet.
